include/zephyr/device/mutable.h

```diff
@@ -0,0 +1,166 @@
+	COND_CODE_1(DT_NODE_EXISTS(node_id), (), (static))                     \
+	STRUCT_SECTION_ITERABLE_NAMED_ALTERNATE(device, device_mutable,                            \
+		Z_DEVICE_SECTION_NAME(level, prio),                            \
+		DEVICE_NAME_GET(dev_id)) =                                     \
+		Z_DEVICE_INIT(name, pm, data, config, api, state, deps)
+
+/**
+ * @brief Define the init entry for a mutable device.
+ *
+ * This macro is similar to @ref Z_DEVICE_INIT_ENTRY_DEFINE but is used for defining
+ * mutable devices.
+ *
+ * @param node_id Devicetree node id for the device (DT_INVALID_NODE if a
+ * software device).
+ * @param dev_id Device identifier.
+ * @param init_fn_ Device init function.
+ * @param level Initialization level.
+ * @param prio Initialization priority.
+ */
+#define Z_DEVICE_INIT_ENTRY_DEFINE_MUT(node_id, dev_id, init_fn_, level, prio)     \
+	static const Z_DECL_ALIGN(struct init_entry) __used __noasan           \
+		Z_INIT_ENTRY_SECTION(level, prio,                              \
+				     Z_DEVICE_INIT_SUB_PRIO(node_id))          \
+		Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(dev_id)) = {                 \
+			.init_fn = {.dev_rw = (init_fn_)},                        \
+			.dev_rw = (struct device *)&DEVICE_NAME_GET(dev_id),                       \
+	}
+
+/**
+ * @brief Define a mutable @ref device and all other required objects.
+ *
+ * This macro is similar to @ref Z_DEVICE_DEFINE but is used for defining
+ * mutable devices.
+ *
+ * @param node_id Devicetree node id for the device (DT_INVALID_NODE if a
+ * software device).
+ * @param dev_id Device identifier (used to name the defined @ref device).
+ * @param name Name of the device.
+ * @param init_fn Device init function.
+ * @param pm Reference to @ref pm_device associated with the device.
+ * (optional).
+ * @param data Reference to device data.
+ * @param config Reference to device config.
+ * @param level Initialization level.
+ * @param prio Initialization priority.
+ * @param api Reference to device API.
+ * @param state Reference to device state.
+ * @param ... Optional dependencies, manually specified.
+ */
+#define Z_DEVICE_DEFINE_MUT(node_id, dev_id, name, init_fn, pm, data, config,      \
+			level, prio, api, state, ...)                          \
+	Z_DEVICE_NAME_CHECK(name);                                             \
+                                                                               \
+	IF_ENABLED(CONFIG_DEVICE_DEPS,                                         \
+		   (Z_DEVICE_DEPS_DEFINE_MUT(node_id, dev_id, __VA_ARGS__);))      \
+                                                                               \
+	Z_DEVICE_BASE_DEFINE_MUT(node_id, dev_id, name, pm, data, config, level,   \
+			     prio, api, state, Z_DEVICE_DEPS_NAME(dev_id));    \
+                                                                               \
+	Z_DEVICE_INIT_ENTRY_DEFINE_MUT(node_id, dev_id, init_fn, level, prio)
```

Comment:
I'm a bit skeptical about all this complexity. Why is this needed? Isn't the dt flag enough?

Reply:
IIRC, this was the minimum required to support 
`DEVICE_DT_INST_DEFINE_MUT()` with a separately named iterable section (`devices_mutable` instead of `devices`). I.e. there was a warning or other error when using the same name, but I can double check.

Also, it kind of makes sense to have a separate iterable sections because the memory for `devices` is in `rodata` (rom) while the memory for `devices_mutable` is in `data` (ram) and they should really be independently contiguous.

Reply:
So are devices forced to use `DEVICE_DT_DEFINE_MUT` to be mutable? I don't think this will scale well, specially because there's an ongoing trend to have class-specific device definition macros (e.g. can, i2c, etc.), and the situation there is already pretty bad in terms of complexity. Isn't a `zephyr,mutable` property enough to decide wether to place the device in a ram section instead of rom? Also note that there's code depending on all devices being on a single list (PM_DEVICE), not sure if this can affect it.

Reply:
Technically speaking we could put logic inside the existing macros. I just wanted to avoid being overly intrusive.

I can respin if that is easier, and then, yeah, it would still use the existing interface.

Reply:
Fixed!